### DISCUSSION TITLE
Unexpected peer cursor position after line break

### DISCUSSION THREAD

**Discussion by Flrande** [toeverything/blocksuite#554] [Q&A] [answered]
When I ran the repository's Playground locally and tried out its collaborative editing feature, I noticed that the focus did not match my expectations in the situation shown in the image below. I haven't yet delved into the code in the repository, but was this the intended result? My guess is that this issue occurs because the Enter operation first deletes the node in the first line and then inserts it in the second line, causing the position of the collaborator to be associated with a deleted data structure and therefore unable to be accurately located.

![image](https://user-images.githubusercontent.com/50035259/208293666-b349d508-9e65-4a16-81b2-d4c98cfb29b7.png)

![image](https://user-images.githubusercontent.com/50035259/208293743-0e111b8b-ce54-4624-b786-0ae05d924b52.png)

![image](https://user-images.githubusercontent.com/50035259/208293843-81e2ab99-9595-4735-8e34-47350ccfb547.png)

**Comment by doodlewind** [comment, marked as answer]
Thanks for your feedback. It's an excellent question about how BlockSuite works!

As a block-based framework, BlockSuite models different blocks into different objects ([Y.Map](https://docs.yjs.dev/api/shared-types/y.map) under the hood). So for paragraphs, each paragraph block holds a different [Y.Text](https://docs.yjs.dev/api/shared-types/y.text) instance. When you perform a line break, we split the paragraph into two paragraphs so they hold different text instances.

After splitting, the right-hand part of the existing paragraph will be deleted. But the peer cursor still references this block. Instead of losing the cursor or throwing an index error, we use the [RelativePosition](https://docs.yjs.dev/api/relative-positions) to transform that previous index to a reasonable one, which is the end of the existing block.

As an example, suppose I'm going to do a line break at 🖍️, and your cursor locates at 🖌️:

```
1234🖍️567890🖌️
```

After the line break, there will be two blocks, and your cursor stays at the first one:

```
1234🖌️

567890
```

That's how the line break and cursor work in BlockSuite for now. 

I would say this is the expected behavior for the current implementation, but as you point out, this could be unclear sometimes. I can keep this issue open for possible improvements we can make here.
